### PR TITLE
Reorg PMEM FreeList: global free entry pool + thread local entry cache

### PR DESCRIPTION
**Pull request by JiayuZzz**
In this patch:

1. Move pmem allocator related codes to directory engine/pmem_allocator
2. Reorg free list:
- Use a global free space entry pool to store free space
- Each write thread has a cache structure, which consists of a active free space list and a backup free space list. Each threads store just freed space to and use free space from own active list. To balance free space entries among threads, if too many entries cached by a thread, newly freed entries will be stored in backup list, then move the whole backup list to entry pool which shared by all threads
- If active list is empty, then swap backup list as active list, or fetch a free list from entry pool

### CODE REVIEW

**Comment by JiayuZzz** [comment]
The move work will be assign to background thread in next patch

**Comment by JiayuZzz** [comment]
@KVKIT

File: engine/pmem_allocator/free_list.cpp [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+
+void FreeList::Push(const SizedSpaceEntry &entry) {
+  space_map_->Set(entry.space_entry.offset, entry.size);
+  auto &thread_cache = thread_cache_[write_thread.id];
+  if (entry.size >= thread_cache.active_entries.size()) {
+    std::lock_guard<SpinMutex> lg(large_entries_spin_);
+    large_entries_.insert(entry);
+  } else {
+    if (thread_cache.active_entries[entry.size].size() >= kMaxCacheEntries) {
+      std::lock_guard<SpinMutex> lg(thread_cache.spins[entry.size]);
```
Comment (KVKIT) [inline engine/pmem_allocator/free_list.cpp:161]: What's the purpose for this lock?

**Comment by KVKIT** [comment]
It's not easy to understand the code, please add an overall description of the design and some comments in code implementation.

**Comment by JiayuZzz** [comment]
> It's not easy to understand the code, please add an overall description of the design and some comments in code implementation.

Added comments for the space entry pool,  I think the complete design should add to the comment after we finish back ground jobs

File: engine/pmem_allocator/free_list.cpp (hunk 1 above)
Comment (JiayuZzz) [reply engine/pmem_allocator/free_list.cpp:161]: for background thread to use

**Comment by KVKIT** [comment]
looks good to me.

**Pull request merged**